### ISSUE TITLE
issue with aggday + red recommitted triangle

### ISSUE THREAD

**Issue by bsoule** [beeminder/road#59]
Turns out this is not a jsbrain-only issue -- it remains when I switch the graph back to python -- but it is incorrect nonetheless.

On my weight goal I derailed on the 28th. I.e., my weight on the 27th was above the road. The recommit stuff picked the correct value for the "recommitted" datapoint and moved the road correctly -- 68kg -- the min of the 27th. However, on the 28th itself, in addition to the weights I registered, withings mis-assigned one of Danny's weights to my account -- the 70.3 that's labeled "ambiguous" in the comments. 

The red triangle that's drawn to indicate the derail is pointing at that datapoint, not the derail value or the "RECOMMITTED" point. (Possibly because it's the max on the 28th? or maybe just because it's the _last_ datapoint on the 28th?).

The red triangle should point at the derailing value (or at the recommit point). Definitely confusing/weird that it's pointing at that 70.3 value. 

<img width="37" alt="weight___b_weight_goal_page" src="https://user-images.githubusercontent.com/46881/52224892-fe98b100-285d-11e9-9740-8775e6dd9a9c.png">

I've put the bb file in the test suite as wrong-recommit-triangle.bb

**Comment by saranli** [comment]
Yes, this is indeed due to the erroneous value being considered as the "worst value" for that day. Here is the relevant code snippet:

      // Adjust derailment markers to indicate worst value for that day
      for (i = 0; i < derails.length; i++) {
        ct = derails[i][0]+bu.SID
        if (worstval.hasOwnProperty(ct)) derails[i][1] = worstval[ct]
      }

Since the erroneous datapoint is specific to this goal and jsbrain has no way of knowing that the value was erroneous, I think this can be safely closed, unless you think we should pick the derailment value rather than the worst value to show for the derailment?

**Comment by bsoule** [comment]
Is it on purpose that we're inventing a new metric ("worst for the day") and using it rather than using the aggday-ed value here? Using the aggday-ed value feels more correct to me. 

Actually, I just double checked and for Do More goals we use the aggday value, not the worst-for-the-day. See here:

<img width="130" alt="elfthing___b_elfthing_goal_page" src="https://user-images.githubusercontent.com/46881/57646148-cbf50780-7574-11e9-8d89-37c6b5d39fb5.png">

**Comment by saranli** [comment]
No idea. I will defer to dreev for this since I faithfully ported from the pybrain code. The relevant computation is as follows:
          let vw = allvals[ct].map(e=>e[0])
          worstval[ct] = (goal.yaw<0)?bu.arrMax(vw):bu.arrMin(vw)
basically computes the min or max for all days with datapoints including (I think), the aggregated value. I agree that this probably needs more thought. Perhaps subscribe dreev to this?

**Comment by dreeves** [comment]
Uluc is right. But there was some reason I went with `worstval` instead of `aggval`. Python code for comparison:

```python
worstval[t] = (max(allvals[t]) if yaw<0 else min(allvals[t]))
```

Fortunately, cmonitor makes this so nice to get to the bottom of! I can try try aggval instead and see what all changes in the testsuite...

**Comment by dreeves** [comment]
https://github.com/beeminder/road/commit/a293ec8b5367234f1e2dc88d0cc871340a992908

**Comment by dreeves** [comment]
To do this really right we need to compute the aggregated datapoint as of the 'RECOMMITTED' datapoint, when the derail happened. http://beeminder.com/changelog#3157
